compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1560,34 +1582,78 @@ impl Statement<'_> {
 /// causing an ICE if they are violated.
 #[derive(Clone, Debug, PartialEq, TyEncodable, TyDecodable, Hash, HashStable, TypeFoldable)]
 pub enum StatementKind<'tcx> {
-    /// Write the RHS Rvalue to the LHS Place.
+    /// Assign statements roughly correspond to an assignment in Rust proper (`x = ...`) except
+    /// without the possibility of dropping the previous value (that must be done separately, if at
+    /// all). The *exact* way this works is undecided. It probably does something like evaluating
+    /// the LHS and RHS, and then doing the inverse of a place to value conversion to write the
+    /// resulting value into memory. Various parts of this may do type specific things that are more
+    /// complicated than simply copying over the bytes depending on the types.
+    ///
+    /// **Needs clarification**: The implication of the above idea would be that assignment implies
+    /// that the resulting value is initialized. I believe we could commit to this separately from
+    /// committing to whatever part of the memory model we would need to decide on to make the above
+    /// paragragh precise. Do we want to?
+    ///
+    /// Assignments in which the types of the place and rvalue differ are not well-formed.
+    ///
+    /// **Needs clarification**: Do we ever want to worry about non-free (in the body) lifetimes for
+    /// the typing requirement in post drop-elaboration MIR? I think probably not - I'm not sure we
+    /// could meaningfully require this anyway. How about free lifetimes? Is ignoring this
+    /// interesting for optimizations? Do we want to allow such optimizations?
+    ///
+    /// **Needs clarification**: We currently require that the LHS place not overlap with any place
+    /// read as part of computation of the RHS. This requirement is under discussion in [#68364]. As
```

Comment:
Note that the discussion makes a distinction between primitives and non-primitives, which is important given that MIR constructed for compound assignment would already violate the non-overlap rule.

Reply:
Can you give an example of this? I didn't find this in the discussion in the issue

Reply:
Without overflow checks:

```rust
pub fn f(mut a: u32) {
    a += 1;
}
```

```rust
// MIR for `f` 0 mir_map

fn f(_1: u32) -> () {
    debug a => _1;                       // in scope 0 at a.rs:1:10: 1:15
    let mut _0: ();                      // return place in scope 0 at a.rs:1:22: 1:22

    bb0: {
        _1 = Add(_1, const 1_u32);       // scope 0 at a.rs:2:5: 2:11
        _0 = const ();                   // scope 0 at a.rs:1:22: 3:2
        return;                          // scope 0 at a.rs:3:2: 3:2
    }
}
```

Reply:
Good point. I don't even know what "those not producing primitives" means.

But having such an already subtle rule only apply sometimes seems terrible. We should either make it apply always, or never, or have a `bool` annotated at the statement that says whether they may overlap.

Reply:
On the one hand MIR building constructs assignments where lhs and rhs are overlapping. On the other hand the LLVM code generation for an assignment uses `llvm.memcpy` which requires destination to be either equal or non-overlapping with source.

Reply:
Well yeah that's bad and fragile and we shouldn't do it IMO. ;)